### PR TITLE
Changes: EC-42 and EC-31

### PR DESCRIPTION
**Pull request by lzhao-skyhook**
1. [EC-42] Create cleaner client interface to Skyhook functionality
2. [EC-31] Scan cache matching on client for stationary detection

### CODE REVIEW

**Comment by lzhao-skyhook** [comment]
@tboinske  - please review.

File: src/client/sky_client.c [hunk 1]
```diff
@@ -34,21 +34,24 @@ an embeddded C project.
...
+
+#include "sky_cache.h"
 #include "sky_protocol.h"
 #include "sky_util.h"
 #include "sky_crypt.h"
 #include "sky_print.h"
 
+#define ENABLE_SCAN_CACHING   // enable client scan caching to eliminate sending location requests when locations remain the same
+
+#define CACHE_MATCH_PERCENTAGE 0.5f
```
Comment (tboinske-skyhook) [inline src/client/sky_client.c]: I suspect that the allowable values for this preprocessor identifier are 0-1. If so, this value does not represent a "percentage" (which by definition takes values between 0 and 100). So I suggest renaming it to CACHE_MATCH_THRESHOLD, and add a comment clarifying it's meaning.

File: src/client/sky_client.c [hunk 2]
```diff
@@ -81,316 +89,284 @@ bool get_ip_addr(struct location_rq_t * p_creq) {
...
+       use the wifi chip MAC address
+       here we just set it to a constant */
+    uint8_t device_mac[] = {0xCA,0xFE,0xBA,0xBE,0xCA,0xFE};
+
+    /* load wifi scan result
+       for demo purposes we assign them manually */
+    const uint8_t ap_count = 4;
+    struct ap_t aps[ap_count];
+    {
+        uint8_t mac[] = {0xe0,0x1c,0x41,0x3b,0xdf,0x29};
```
Comment (tboinske-skyhook) [inline src/client/sky_client.c]: Be consistent with your whitespace convention - line 265 includes a single space following each comma (which increases readability). I suggest doing the same on line 264 (and its brethren below).

File: src/client/sky_client.c [hunk 3]
```diff
@@ -81,316 +89,284 @@ bool get_ip_addr(struct location_rq_t * p_creq) {
...
+//        exit(-1);
+//    }
+//
+//    memcpy(&resp.key, &key, sizeof(resp.key));
+//
+//    cnt = sky_recv_location_response(&resp, rx, &sockfd);
+//    if (cnt < 0) {
+//        perror("Failed to receive location response\n");
+//        exit(-1);
+//    }
```
Comment (tboinske-skyhook) [inline src/client/sky_client.c]: Why leave this commented code in place?

File: src/client/sky_client.c [hunk 4]
```diff
@@ -81,316 +89,284 @@ bool get_ip_addr(struct location_rq_t * p_creq) {
...
+#endif
 
     // print location response
     puts("--------------");
     puts("PARSED RESPONSE");
     puts("--------------");
     print_location_resp(&rsp);
 
+#ifdef ENABLE_SCAN_CACHING
+    sky_cache_deinit();
```
Comment (tboinske-skyhook) [inline src/client/sky_client.c]: I assume "deinit" is short for "deinitialize". However, "deinitialize" is not an actual word. I suggest "clean_up" instead.

But in any event, as noted above, why does the application code need to call this function, or any caching-related functions?

File: src/client/sky_client.c [hunk 5]
```diff
@@ -81,316 +89,284 @@ bool get_ip_addr(struct location_rq_t * p_creq) {
...
+}
+
+//
+// Entry function
+//
+
+int main(int argc, char**argv) {
+
+#ifdef ENABLE_SCAN_CACHING
+    sky_cache_init(MAC_SIZE, sizeof(int8_t));
```
Comment (tboinske-skyhook) [inline src/client/sky_client.c]: I thought that we agreed that all of the caching machinery would be hidden inside sky_query_location()?

**Comment by tboinske-skyhook** [comment]
Also - your comment message doesn't adhere to [the convention that we've discussed in the past](http://chris.beams.io/posts/git-commit/).

File: src/client/sky_client.c (hunk 1 above)
Comment (lzhao-skyhook) [inline src/client/sky_client.c]: Threshold may also be integer. Renamed to CACHE_MATCH_RATIO. Please let me know if it is still not clear.

File: src/client/sky_client.c (hunk 5 above)
Comment (lzhao-skyhook) [inline src/client/sky_client.c]: Sorry for the mis-sync-up. I have updated my reply in wiki yesterday @ https://skyhookwireless.atlassian.net/wiki/display/ELG/Design+of+Scan+Cache+Matching+-+Stationary+Detection?focusedCommentId=95704090#comment-95704090. Thought you would be notified automatically... I will continue the discussion with you here.

Below is my thoughts:

After diving into implementation, it looks more naturally fit to leave check_cache_match() outside of sky_query_location(). Reasons are below:
- sky_query_location() is just one query without context. Even if detecting the location remains the same, sky_query_location() does not cache the earlier response to return.
- sky_protocol.c focused on client ELGv2 protocol APIs. Caching is not naturally fit into protocol, and protocol is not supposed to put dependency on cache neither.
- Caching naturally fit into client's code as the context is hold there. The cache code and protocol code is completely independent to each other. So caching belongs to client helper package.

Further, the macro #ENABLE_SCAN_CACHING is not as useful as I thought before. The cache implementation code cannot be skipped as GCC requires it be a linking library. So the cache implementation code will be compiled regardless. It just matters if elg_client decides to link it or not.

File: src/client/sky_client.c (hunk 2 above)
Comment (lzhao-skyhook) [inline src/client/sky_client.c]: Done. I thought byte array is usually long, so typically I do not put spaces to separate byte values. Here is the mac address, so it is ok to put spaces in.

File: src/client/sky_client.c (hunk 3 above)
Comment (lzhao-skyhook) [inline src/client/sky_client.c]: There are two approaches for function calls: blocking and non-blocking. Both are provided by sky_protocol.c for satisfying different use cases. The adopted is blocking call. The commented lines 341-353 are for non-blocking calls. This is commented at lines 330-332.

File: src/client/sky_client.c (hunk 4 above)
Comment (lzhao-skyhook) [inline src/client/sky_client.c]: sky_cache_init() and sky_cache_deinit() are introduced in pair. They are used to allocate and free heap memory for cache. The lifetime of cache is determined out of the context of sky_query_location(). It does not make a lot of sense to introduce sky_multi_query_locations(). Scanning networks and querying locations in loop belong to sky_client.c. Cache is applied in the loop. So is the init and deinit of cache. Hence, the cache code is pulled out of sky_query_location().

**Comment by lzhao-skyhook** [comment]
I will add another commit just for summarizing the code changes.

**Pull request merged**